File: cs50/sentimental-readability/readability.py

def get_letters(input_text):

    letters = sum(1 for char in input_text if char.isalpha())
    return letters

def get_words(input_text):
    words = sum(1 for char in input_text if char == ' ') + 1
    return words

def get_sentences(input_text):
    sentences = sum(1 for char in input_text if char in '.!?')
    return sentences

def coleman_liau_idx(letters, words, sentences):

    L = (letters / words) * 100
    S = (sentences / words) * 100
    index = round(0.0588 * L - 0.296 * S - 15.8)
    return index

def main():

    input_text = input("Text: ")

    letters = get_letters(input_text)
    words = get_words(input_text)
    sentences = get_sentences(input_text)

    index = coleman_liau_idx(letters, words, sentences)

    if index < 1:
        print("Before Grade 1")
    elif index > 16:
        print("Grade 16+")
    else:
        print(f"Grade {index}")


main()

File: cs50/readability/readability.c
#include <cs50.h>
#include <math.h>

#include <ctype.h>
#include <stdio.h>

#include <stdlib.h>
#include <string.h>

int get_letters(  char input_text[], int len);
int get_words(    char input_text[], int len);
int get_sentences(char input_text[], int len);
int ColemanLiau_Idx(int letters, int words, int sentences);

int main (void)
{
    string input_text = get_string("Text: ");
    int length        = strlen(input_text);

    int letters   = get_letters(  input_text, length);
    int words     = get_words(    input_text, length);
    int sentences = get_sentences(input_text, length);

    int index     = ColemanLiau_Idx(letters, words, sentences);

    if (index < 1)
        printf("Before Grade 1\n");

    else if(index > 16)
        printf("Grade 16+\n");

    else
        printf("Grade %i\n", index);
}

int get_letters(char input_text[], int len)
{
    int letters = 0;
    for (int i = 0; i < len; i++)
    {
        if ((input_text[i] >= 'a' && input_text[i] <= 'z') || (input_text[i] >= 'A' && input_text[i] <= 'Z'))
        {
            letters++;
        }
    }
    return letters;
}

int get_words(char input_text[], int len)
{
    int words = 0;
    for (int i = 0; i < len; i++)
    {
        if (input_text[i] == ' ')
        {
            words++;
        }
    }
    // + 1 to count the last word, assuming text isn't empty
    return words + 1;
}

int get_sentences(char input_text[], int len)
{
    int sentences = 0;
    for (int i = 0; i < len; i++)
    {
        if (input_text[i] == '.' || input_text[i] == '!' || input_text[i] == '?')
        {
            sentences++;
        }
    }
    return sentences;
}

int ColemanLiau_Idx(int letters, int words, int sentences)
{
    float L = ((float)letters / words) * 100;
    float S = ((float)sentences / words) * 100;
    int index = round(0.0588 * L - 0.296 * S - 15.8);
    return index;
}

